README.md
GeeXboX uShare (Modified Version)
================================================

GeeXboX uShare is a UPnP (TM) A/V Media Server. It implements the server
component that provides UPnP media devices with information on available
multimedia files. uShare uses the built-in http server of libupnp to
stream the files to clients.

GeeXboX uShare is able to provide access to both images, videos, music
or playlists files (see below for a complete file format support list).
It does not act as an UPnP Media Adaptor and thus, can't transcode
streams to fit the client requirements.

uShare is written in C for the GeeXboX project (see http://www.geexbox.org/).
It is designed to provide access to multimedia contents to GeeXboX but can of
course be used by any other UPnP client device.

This is a modified version of the original release. Including some bug fixes and modifications in order to compile and run on **FreeBSD**. See *ChangeLog* for details. For the remaining part of the document, if not explicitly speciifed, it's assumed that the underlying os is a FreeBSD system. Some additional settings  might be necessary if it would be used for linux systems. See section **For Linux Users**.

GeeXboX uShare is free software - it is licensed under the terms of the GNU
General Public License (GPL).

Homepage (the original release)
================================

Web site and file area for uShare is hosted on GeeXboX server :

   http://ushare.geexbox.org/

The latest version of uShare should always be available on this site.

Requirements
============

The following programs are required to build uShare:

 * GNU Make

   GNU Make is a tool which controls the generation of executables and other non-source files of a program from the program's source files. (see https://www.gnu.org/software/make/)

 * pkg-config.

   pkg-config is a helper tool used when compiling applications and libraries.
   It helps you insert the correct compiler options on the command line.
   (see http://pkg-config.freedesktop.org/wiki/ ).

The following UPnP library is required to build and run uShare :

 * libupnp, 1.4.2 or later

   The libupnp library is used to communicate using the UPnP protocol.
   libupnp can be downloaded from http://pupnp.sourceforge.net/.

The following DLNA library is required for proper DLNA support :

 * libdlna, 0.2.1 or later

   The libdlna library is used to provides DLNA profiles informations.
   libdlna can be downloaded from http://libdlna.geexbox.org/.

Building and Installation
=========================

Compile uShare by running configure and then make. This should
produce an executable ushare in the src subdirectory, which can be
used right away. No extra files need to be installed.

You can pass the CFLAGS you want to configure including -DDEBUG in order
to activate support for debug messages in uShare.

Example :
```shell
CFLAGS="-Os" ./configure --prefix=/usr
gmake
```

You can enable DLNA support by doing:
```shell
./configure --enable-dlna
```

If you want to install uShare on your system, run :
```shell
gmake install-strip
```

This will copy the executable and manual page into their appropriate
directories (/usr/bin and /usr/man/man1 in this example).

Usage
=====

uShare runs from the console only. It supports the usual --help option
which displays usage and option information.

```shell
Usage: ushare [-n name] [-i interface] [-c directory] [[-c directory] ...]
Options:
 -n, --name=NAME        Set UPnP Friendly Name (default is 'uShare')
 -i, --interface=IFACE  Use IFACE Network Interface (default is 'eth0')
 -f, --cfg=FILE         Config file to be used
 -p, --port=PORT        Forces the HTTP server to run on PORT
 -q, --telnet-port=PORT Forces the TELNET server to run on PORT
 -c, --content=DIR      Share the content of DIR directory.
 -w, --no-web           Disable the control web page (enabled by default)
 -t, --no-telnet        Disable the TELNET control (enabled by default)
 -o, --override-iconv-err       If iconv fails parsing name, still add to media contents (hoping the renderer can handle it)
 -v, --verbose          Set verbose display
 -x, --xbox             Use XboX 360 compliant profile
 -d, --dlna             Use DLNA compliant profile (PlayStation3 needs this)
 -D, --daemon           Run as a daemon.
 -V, --version          Display the version of uShare and exit
 -h, --help             Display this help
```

uShare gets its configuration from the `/etc/ushare.conf` file.
You can force configuration options through command line.

uShare expects one or several directory argument (-c argument),
specifying where multimedia files are stored. You should probably also use
the -i option to specify which interface uShare should listen on.

```shell
ushare -c /shares
ushare -c /shares1 --content=/shares2
```

You can also perform remote control of uShare UPnP Media Server through its
web interface. This let you define new content locations at runtime or
update the currently shared one in case the filesystem has changed.
Just go to : `http://ip_address:port/web/ushare.html`

See the manual page for more details :

```shell
man ushare
```

Supported File Formats List
===========================

- Video files :
  asf, avi, dv, divx, wmv, mjpg, mjpeg, mpeg, mpg, mpe, mp2p, vob, mp2t, m1v, m2v, m4v, m4p, mp4ps, ts, ogm, mkv, rmvb, mov, qt

- Audio files :
  aac, ac3, aif, aiff, at3p, au, snd, dts, rmi, mp1, mp2, mp3, mp4, mpa, ogg, wav, pcm, lpcm, l16, wma, mka, ra, rm, ram

- Images files :
  bmp, ico, gif, jpeg, jpg, jpe, pcd, png, pnm, ppm, qti, qtf, qtif, tif, tiff

- Playlist files :
  pls, m3u, asx

If you want uShare to support more file formats, simply add its properties
in the src/mime.c table. Do not forget to send a patch to update uShare.

Known bugs and limitations
==========================

If you need to listen on more than one interface, you will have to start
multiple instances of the media server.

uShare keeps some information on files in memory.
If your multimedia collection is huge, this might be a problem.

For Linux Users
===============

The source code should just work under any Linux os as it does under FreeBSD systems. But the default settings in `configure` script is meant to work only for FreeBSD, Linux users need to speficy the compiler (via `CC`) and make system (via `MAKE`) accordingly.In the usual case, Gcc is the default comiler and gnu make is the default make system on linux.

For example:
```shell
CC=gcc MAKE=make ./configure ...
```

Also, since the `make` command on linux is default to be gnu make, all the `gmake` commands mentioned above should be replaced with `make`.

Trademarks
==========

UPnP(TM) is a trademark of the UPnP(TM) Implementers Corporation.


ChangeLog
2018-02-04  Junsheng Guo
        * fixed build for FreeBSD.
	* include necessary c header files to avoid compiler warning and seg fault
	* use setlocale to get locale settings from current environment

2007-12-09  Benjamin Zores <ben@geexbox.org>

	* new public release (Version 1.1a).

2007-12-04  Benjamin Zores <ben@geexbox.org>

	* new public release (Version 1.1).

2007-11-04  Benjamin Zores <ben@geexbox.org>

	* much complete DLNA support through libdlna.

2007-07-30  Benjamin Zores <ben@geexbox.org>

	* MacOSX build fix.

2007-07-05  Alexis Saettler <asbin@asbin.org>

	* new public release (Version 1.0).

2007-06-26  Benjamin Zores  <ben@geexbox.org>

	* fix for XboX 360 file discovery (patch by Keith Brindley).

2007-06-25  Benjamin Zores  <ben@geexbox.org>

	* added support for DLNA (Digital Living Network Alliance),
	  the wonderful closed-source marketing UPnP++ specification.
	  uShare is now compliant with PlayStation 3 UPnP/DLNA Media Renderer.

2007-02-26  Alexis Saettler <asbin@asbin.org>

	* new public release (Version 0.9.10).
	* fixed http 404 error which prevented ushare 0.9.9 release to work.

2007-02-25  Alexis Saettler <asbin@asbin.org>

	* new public release (Version 0.9.9).
	* fixed bug with localization, use now --localedir configure's
	  option to set location of .mo files.

2007-01-23  Benjamin Zores  <ben@geexbox.org>

	* make libupnp >= 1.4.2 a requirement.

2007-01-20  Benjamin Zores  <ben@geexbox.org>

	* added support for FreeBSD.

2007-01-19  Benjamin Zores  <ben@geexbox.org>

	* use a RedBlack balanced tree searching and sorting algorithm
	  instead of the recursive one. It avoids recursive calls, takes
	  less memory (perfect for embedded systems) and executes in more
	  "constant time" than the logarithmic style.
	  Original patch by Phil Chandler.	
	* massive memory optimization (uShare now takes 75% less RAM ;-)
	  only not using fixed size URL allocation buffer (a 16000+ files
	  collection now only takes 5MB where it used to take 19MB).
	* improved search parsing and matching mechanism (patch by Mostafa
	  Hosseini). XboX360 support is now complete.
	  Please note that to stream video to the XboX, the console should
	  be running dashboard version 2.0.4552.0 and only unprotected
	  .wmv files are supported.
	
2007-01-14  Benjamin Zores  <ben@geexbox.org>

	* make libupnp >= 1.4.1 a requirement.
	* fixed support for 2GB+ files (see http://sourceforge.net/tracker/index.php?func=detail&aid=1634927&group_id=151880&atid=782410)
	* fixed support for XboX 360 (thanks to Mostafa Hosseini).
	* added basic support for 'Search' action.
	
2006-12-10  Alexis Saettler <asbin@asbin.org>

	* new public release (Version 0.9.8).

2006-10-19  Alexis Saettler <asbin@asbin.org>

	* add German translation (thanks to Robin H.).
	* fix filtering and browse capabilities : now uShare is conform to
	ContentDirectory:1 Service Template Standard.

2006-09-17  Alexis Saettler <asbin@asbin.org>

	* add iso and bin mime types.
	* convert special xml characters to xml (like & => &amp;)
	  (reported by Frank Scholz).

2006-09-10  Alexis Saettler <asbin@asbin.org>

	* Use sysconfdir (from configure) to set /etc destination
	* update init script to use lsb functions

2006-03-12  Benjamin Zores  <ben@geexbox.org>

	* new public release (Version 0.9.7)

2006-03-07  Benjamin Zores  <ben@geexbox.org>

	* modify configure.ac to check for functions in libupnp that
	  only comes with releases >= 1.3.1
	  (previous versions of the library are buggy and should not be used).
	* added new Microsoft Registrar service which fakes the authorization
	  required to be able to stream to MS compliant Media Renderers
	  (i.e now usable with XboX 360 and probably others).
	
2006-02-27  Benjamin Zores  <ben@geexbox.org>

	* applied suggestions from Julien Lincy
	* UPnP CDS compliance fix when request count and index are 0.
	* removed the ':' characters from UDN anming convention (devices like
	  the Roku SoundBridge M2000 Network Music Player do not support it).

2006-02-19  Benjamin Zores  <ben@geexbox.org>

	* new public release (Version 0.9.6)

2006-01-06  Benjamin Zores  <ben@geexbox.org>

	* added Fedora RPM build file (ushare.spec) and startup script.
	  (patch by Eric Tanguy <eric.tanguy@univ-nantes.fr>)

2005-12-28  Benjamin Zores  <ben@geexbox.org>

	* can now specify on which port the HTTP server has to run.
	  (useful for security and firewall issues).

2005-12-27  Benjamin Zores  <ben@geexbox.org>

	* fixed a potential buffer overrun when building DIDL messages.
	  (patch by Bernd Loske <bernd@loskeonline.net> ).

2005-12-19  Benjamin Zores  <ben@geexbox.org>

	* new public release (Version 0.9.5)

2005-12-17  Benjamin Zores  <ben@geexbox.org>

	* added UPnP presentation URL support.
	  (you can now update or add contents through a web interface).

2005-12-10  Benjamin Zores  <ben@geexbox.org>

	* added support for subtitle file formats.
	* added support for DVD specific file formats.

2005-12-08  Benjamin Zores  <ben@geexbox.org>

	* new public bug-fix release (Version 0.9.4)
	* fixed some memleaks.
	* avoid having empty shared directory name
  	  (crash with djmount, thanks to Remi Turboult for reporting it).
	* fixed handling of non-absolute content directories paths.

2005-11-23  Alexis Saettler  <asbin@asbin.org>

	* add logging with syslog support, in daemon mode.

2005-11-20  Benjamin Zores  <ben@geexbox.org>

	* new public bug-fix release (Version 0.9.3)
	* Added verbose mode.
	* Read configuration from /etc/ushare.conf configuration file.
	* Support for daemon mode.
	* Better MIME types handling.
	* Support for new multimedia files extensions.
	* Rewrite of some string buffer and integer manipulation tools.

2005-11-16  Alexis Saettler  <asbin@asbin.org>

	* modify -c parameter habdle, to allow multiple shared directories.

2005-11-13  Benjamin Zores  <ben@geexbox.org>

	* new public bug-fix release (Version 0.9.2)

2005-11-13  Benjamin Zores  <ben@geexbox.org>


src/ushare.h
/*
 * ushare.h : GeeXboX uShare UPnP Media Server header.
 * Originally developped for the GeeXboX project.
 * Parts of the code are originated from GMediaServer from Oskar Liljeblad.
 * Copyright (C) 2005-2007 Benjamin Zores <ben@geexbox.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#ifndef _USHARE_H_
#define _USHARE_H_

#include <upnp/upnp.h>
#include <upnp/upnptools.h>
#include <stdbool.h>
#include <pthread.h>

#ifdef HAVE_DLNA
#include <dlna.h>
#endif /* HAVE_DLNA */

#include "content.h"
#include "buffer.h"
#include "redblack.h"

#define VIRTUAL_DIR "/web"
#define XBOX_MODEL_NAME "Windows Media Connect Compatible"
#define DEFAULT_UUID "898f9738-d930-4db4-a3cf"

#define UPNP_MAX_CONTENT_LENGTH 4096

#define STARTING_ENTRY_ID_DEFAULT 0
#define STARTING_ENTRY_ID_XBOX360 100000

#define UPNP_DESCRIPTION \
"<?xml version=\"1.0\" encoding=\"utf-8\"?>" \
"<root xmlns=\"urn:schemas-upnp-org:device-1-0\">" \
"  <specVersion>" \
"    <major>1</major>" \
"    <minor>0</minor>" \
"  </specVersion>" \
"  <device>" \
"    <deviceType>urn:schemas-upnp-org:device:MediaServer:1</deviceType>" \
"    <friendlyName>%s: 1</friendlyName>" \
"    <manufacturer>GeeXboX Team</manufacturer>" \
"    <manufacturerURL>http://ushare.geexbox.org/</manufacturerURL>" \
"    <modelDescription>GeeXboX uShare : UPnP Media Server</modelDescription>" \
"    <modelName>%s</modelName>" \
"    <modelNumber>001</modelNumber>" \
"    <modelURL>http://ushare.geexbox.org/</modelURL>" \
"    <serialNumber>GEEXBOX-USHARE-01</serialNumber>" \
"    <UDN>uuid:%s</UDN>" \
"    <serviceList>" \
"      <service>" \
"        <serviceType>urn:schemas-upnp-org:service:ConnectionManager:1</serviceType>" \
"        <serviceId>urn:upnp-org:serviceId:ConnectionManager</serviceId>" \
"        <SCPDURL>/web/cms.xml</SCPDURL>" \
"        <controlURL>/web/cms_control</controlURL>" \
"        <eventSubURL>/web/cms_event</eventSubURL>" \
"      </service>" \
"      <service>" \
"        <serviceType>urn:schemas-upnp-org:service:ContentDirectory:1</serviceType>" \
"        <serviceId>urn:upnp-org:serviceId:ContentDirectory</serviceId>" \
"        <SCPDURL>/web/cds.xml</SCPDURL>" \
"        <controlURL>/web/cds_control</controlURL>" \
"        <eventSubURL>/web/cds_event</eventSubURL>" \
"      </service>" \
"      <service>" \
"        <serviceType>urn:microsoft.com:service:X_MS_MediaReceiverRegistrar:1</serviceType>\n" \
"        <serviceId>urn:microsoft.com:serviceId:X_MS_MediaReceiverRegistrar</serviceId>\n" \
"        <SCPDURL>/web/msr.xml</SCPDURL>" \
"        <controlURL>/web/msr_control</controlURL>" \
"        <eventSubURL>/web/msr_event</eventSubURL>" \
"      </service>\n" \
"    </serviceList>" \
"    <presentationURL>/web/ushare.html</presentationURL>" \
"  </device>" \
"</root>"

struct ushare_t {
  char *name;
  char *interface;
  char *model_name;
  content_list *contentlist;
  struct rbtree *rb;
  struct upnp_entry_t *root_entry;
  int nr_entries;
  int starting_id;
  int init;
  UpnpDevice_Handle dev;
  char *udn;
  char *ip;
  unsigned short port;
  unsigned short telnet_port;
  struct buffer_t *presentation;
  bool use_presentation;
  bool use_telnet;
#ifdef HAVE_DLNA
  bool dlna_enabled;
  dlna_t *dlna;
  dlna_org_flags_t dlna_flags;
#endif /* HAVE_DLNA */
  bool xbox360;
  bool verbose;
  bool daemon;
  bool override_iconv_err;
  char *cfg_file;
  pthread_mutex_t termination_mutex;
  pthread_cond_t termination_cond;
};

struct action_event_t {
  struct Upnp_Action_Request *request;
  bool status;
  struct service_t *service;
};

void display_headers ();

#endif /* _USHARE_H_ */


src/services.c
/*
 * services.c : GeeXboX uShare UPnP services handler.
 * Originally developped for the GeeXboX project.
 * Parts of the code are originated from GMediaServer from Oskar Liljeblad.
 * Copyright (C) 2005-2007 Benjamin Zores <ben@geexbox.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <stdlib.h>
#include <string.h>
#include <upnp/upnp.h>
#include <upnp/upnptools.h>

#include "ushare.h"
#include "services.h"
#include "cms.h"
#include "cds.h"
#include "msr.h"
#include "trace.h"

/* Represent the ObjectID argument. */
#define ARG_OBJECT_ID "ObjectID"

/* Represent the ContainerID argument. */
#define ARG_CONTAINER_ID "ContainerID"

extern struct service_action_t cds_service_actions[];
extern struct service_action_t cms_service_actions[];
extern struct service_action_t msr_service_actions[];

static struct service_t services[] = {
  {
    CDS_SERVICE_ID,
    CDS_SERVICE_TYPE,
    cds_service_actions
  },
  {
    CMS_SERVICE_ID,
    CMS_SERVICE_TYPE,
    cms_service_actions
  },
  {
    MSR_SERVICE_ID,
    MSR_SERVICE_TYPE,
    msr_service_actions
  },
  { NULL, NULL, NULL }
};

bool
find_service_action (struct Upnp_Action_Request *request,
                     struct service_t **service,
                     struct service_action_t **action)
{
  int c, d;

  *service = NULL;
  *action = NULL;

  if (!request) {
    return false;
  }

  for (c = 0; services[c].id != NULL; c++)
    if (!strcmp (services[c].id, request->ServiceID))
    {
      *service = &services[c];
      for (d = 0; services[c].actions[d].name; d++)
      {
        if (!strcmp (services[c].actions[d].name, request->ActionName))
        {
          *action = &services[c].actions[d];
          return true;
        }
      }
      return false;
    }

  return false;
}

bool
upnp_add_response (struct action_event_t *event, char *key, const char *value)
{
  char *val;
  int res;

  if (!event || !event->status || !key || !value)
    return false;

  val = strdup (value);
  if (!val)
    return false;

  res = UpnpAddToActionResponse (&event->request->ActionResult,
                                 event->request->ActionName,
                                 event->service->type, key, val);

  if (res != UPNP_E_SUCCESS)
    {
      free (val);
      return false;
    }

  free (val);
  return true;
}

char *
upnp_get_string (struct Upnp_Action_Request *request, const char *key)
{
  IXML_Node *node = NULL;

  if (!request || !request->ActionRequest || !key)
    return NULL;

  node = (IXML_Node *) request->ActionRequest;
  if (!node)
  {
    log_verbose ("Invalid action request document\n");
    return NULL;
  }

  node = ixmlNode_getFirstChild (node);
  if (!node)
  {
    log_verbose ("Invalid action request document\n");
    return NULL;
  }

  node = ixmlNode_getFirstChild (node);
  for (; node; node = ixmlNode_getNextSibling (node))
    if (!strcmp (ixmlNode_getNodeName (node), key))
    {
      node = ixmlNode_getFirstChild (node);
      if (!node)
        return strdup ("");
      return strdup (ixmlNode_getNodeValue (node));
    }

  log_verbose ("Missing action request argument (%s)\n", key);

  return NULL;
}

int
upnp_get_ui4 (struct Upnp_Action_Request *request, const char *key)
{
  char *value;
  int val;

  if (!request || !key)
    return 0;

  value = upnp_get_string (request, key);
  if (!value && !strcmp (key, ARG_OBJECT_ID))
    value = upnp_get_string (request, ARG_CONTAINER_ID);

  if (!value)
    return 0;

  val = atoi (value);
  free (value);

  return val;
}


src/util_iconv.c
/*
 * util_iconv.c : GeeXboX uShare iconv string encoding utlities.
 * Originally developped for the GeeXboX project.
 * Copyright (C) 2005-2007 Alexis Saettler <asbin@asbin.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "util_iconv.h"

#if HAVE_ICONV
#include <iconv.h>
static iconv_t cd = 0;
#endif

#if HAVE_LANGINFO_CODESET
#include <locale.h>
#endif

void
setup_iconv (void)
{

#if HAVE_ICONV && HAVE_LANGINFO_CODESET
  char *mycodeset = NULL, *native_env_locale = NULL;
  native_env_locale = setlocale(LC_ALL, "");

  mycodeset = strchr(native_env_locale, '.');
  
  if (!mycodeset) {
    return;
  } else {
    ++mycodeset;
  }
  

  /**
   * Setup conversion descriptor if user's console is non-UTF-8. Otherwise
   * we can just leave cd as NULL
   */
  if (strcmp (mycodeset, UTF8))
  {
    cd = iconv_open (UTF8, mycodeset);
    if (cd == (iconv_t) (-1))
    {
      perror ("iconv_open");
      cd = 0;
    }
  }
#endif

}

void
finish_iconv (void)
{
#if HAVE_ICONV
  if (!cd)
    return;
  if (iconv_close (cd) < 0)
    perror ("iconv_close");
  cd = 0;
#endif
}

/**
 * iconv_convert : convert a string, using the current codeset
 * return: a malloc'd string with the converted result
 */
char *
iconv_convert (const char *input)
{
#if HAVE_ICONV
  size_t inputsize = strlen (input) + 1;
  size_t dummy = 0;
  size_t length = 0;
  char *result;
  char *inptr, *outptr;
  size_t insize, outsize;

  /* conversion not necessary. save our time. */
  if (!cd)
    return strdup (input);

  /* Determine the length we need. */
  iconv (cd, NULL, NULL, NULL, &dummy);
  {
    static char tmpbuf[BUFSIZ];
    inptr = (char*) input;
    insize = inputsize;
    while (insize > 0)
    {
      outptr = tmpbuf;
      outsize = BUFSIZ;
      if (iconv (cd, &inptr, &insize, &outptr, &outsize) == (size_t) (-1))
      {
        /**
         * if error is EINVAL or EILSEQ, conversion must be stoped,
         * but if it is E2BIG (not enough space in buffer), we just loop again
         */
        if( errno != E2BIG)
        {
          perror ("error iconv");
          return NULL;
        }
      }
      length += outptr - tmpbuf;
    }

    outptr = tmpbuf;
    outsize = BUFSIZ;
    if (iconv (cd, NULL, NULL, &outptr, &outsize) == (size_t) (-1))
    {
      perror ("error iconv");
      return NULL;
    }
    length += outptr - tmpbuf;
  }

  /* length determined, allocate result space */
  if ((result = (char*) malloc (length * sizeof (char))) == NULL)
  {
    perror ("error malloc");
    return NULL;
  }

  /* Do the conversion for real. */
  iconv (cd, NULL, NULL, NULL, &dummy);
  {
    inptr = (char*) input;
    insize = inputsize;
    outptr = result;
    outsize = length;
    while (insize > 0)
    {
      if (iconv (cd, &inptr, &insize, &outptr, &outsize) == (size_t) (-1))
      {
        if (errno != E2BIG)
        {
          perror ("error iconv");
          free (result);
          return NULL;
        }
      }
    }
    if (iconv (cd, NULL, NULL, &outptr, &outsize) == (size_t) (-1))
    {
      perror ("error iconv");
      free (result);
      return NULL;
    }

    if (outsize != 0)
      abort ();
  }

  return result;
#else
  return strdup (input);
#endif
}


src/mime.c
/*
 * mime.c : GeeXboX uShare media file MIME-type association.
 * Originally developped for the GeeXboX project.
 * Ref : http://freedesktop.org/wiki/Standards_2fshared_2dmime_2dinfo_2dspec
 * Copyright (C) 2005-2007 Benjamin Zores <ben@geexbox.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "mime.h"
#include "ushare.h"

#define UPNP_VIDEO "object.item.videoItem"
#define UPNP_AUDIO "object.item.audioItem.musicTrack"
#define UPNP_PHOTO "object.item.imageItem.photo"
#define UPNP_PLAYLIST "object.item.playlistItem"
#define UPNP_TEXT "object.item.textItem"

const struct mime_type_t MIME_Type_List[] = {
  /* Video files */
  { "asf",   UPNP_VIDEO, "http-get:*:video/x-ms-asf:"},
  { "avc",   UPNP_VIDEO, "http-get:*:video/avi:"},
  { "avi",   UPNP_VIDEO, "http-get:*:video/avi:"},
  { "dv",    UPNP_VIDEO, "http-get:*:video/x-dv:"},
  { "divx",  UPNP_VIDEO, "http-get:*:video/avi:"},
  { "wmv",   UPNP_VIDEO, "http-get:*:video/x-ms-wmv:"},
  { "mjpg",  UPNP_VIDEO, "http-get:*:video/x-motion-jpeg:"},
  { "mjpeg", UPNP_VIDEO, "http-get:*:video/x-motion-jpeg:"},
  { "mpeg",  UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mpg",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mpe",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mp2p",  UPNP_VIDEO, "http-get:*:video/mp2p:"},
  { "vob",   UPNP_VIDEO, "http-get:*:video/mp2p:"},
  { "mp2t",  UPNP_VIDEO, "http-get:*:video/mp2t:"},
  { "m1v",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "m2v",   UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "mpg2",  UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "mpeg2", UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "m4v",   UPNP_VIDEO, "http-get:*:video/mp4:"},
  { "m4p",   UPNP_VIDEO, "http-get:*:video/mp4:"},
  { "mp4ps", UPNP_VIDEO, "http-get:*:video/x-nerodigital-ps:"},
  { "ts",    UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "ogm",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mkv",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "rmvb",  UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mov",   UPNP_VIDEO, "http-get:*:video/quicktime:"},
  { "hdmov", UPNP_VIDEO, "http-get:*:video/quicktime:"},
  { "qt",    UPNP_VIDEO, "http-get:*:video/quicktime:"},
  { "bin",   UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "iso",   UPNP_VIDEO, "http-get:*:video/mpeg2:"},

  /* Audio files */
  { "3gp",  UPNP_AUDIO, "http-get:*:audio/3gpp:"},
  { "aac",  UPNP_AUDIO, "http-get:*:audio/x-aac:"},
  { "ac3",  UPNP_AUDIO, "http-get:*:audio/x-ac3:"},
  { "aif",  UPNP_AUDIO, "http-get:*:audio/aiff:"},
  { "aiff", UPNP_AUDIO, "http-get:*:audio/aiff:"},
  { "at3p", UPNP_AUDIO, "http-get:*:audio/x-atrac3:"},
  { "au",   UPNP_AUDIO, "http-get:*:audio/basic:"},
  { "snd",  UPNP_AUDIO, "http-get:*:audio/basic:"},
  { "dts",  UPNP_AUDIO, "http-get:*:audio/x-dts:"},
  { "rmi",  UPNP_AUDIO, "http-get:*:audio/midi:"},
  { "mid",  UPNP_AUDIO, "http-get:*:audio/midi:"},
  { "mp1",  UPNP_AUDIO, "http-get:*:audio/mp1:"},
  { "mp2",  UPNP_AUDIO, "http-get:*:audio/mp2:"},
  { "mp3",  UPNP_AUDIO, "http-get:*:audio/mpeg:"},
  { "mp4",  UPNP_AUDIO, "http-get:*:audio/mp4:"},
  { "m4a",  UPNP_AUDIO, "http-get:*:audio/mp4:"},
  { "ogg",  UPNP_AUDIO, "http-get:*:audio/x-ogg:"},
  { "wav",  UPNP_AUDIO, "http-get:*:audio/wav:"},
  { "pcm",  UPNP_AUDIO, "http-get:*:audio/l16:"},
  { "lpcm", UPNP_AUDIO, "http-get:*:audio/l16:"},
  { "l16",  UPNP_AUDIO, "http-get:*:audio/l16:"},
  { "wma",  UPNP_AUDIO, "http-get:*:audio/x-ms-wma:"},
  { "mka",  UPNP_AUDIO, "http-get:*:audio/mpeg:"},
  { "ra",   UPNP_AUDIO, "http-get:*:audio/x-pn-realaudio:"},
  { "rm",   UPNP_AUDIO, "http-get:*:audio/x-pn-realaudio:"},
  { "ram",  UPNP_AUDIO, "http-get:*:audio/x-pn-realaudio:"},
  { "flac", UPNP_AUDIO, "http-get:*:audio/x-flac:"},

  /* Images files */
  { "bmp",  UPNP_PHOTO, "http-get:*:image/bmp:"},
  { "ico",  UPNP_PHOTO, "http-get:*:image/x-icon:"},
  { "gif",  UPNP_PHOTO, "http-get:*:image/gif:"},
  { "jpeg", UPNP_PHOTO, "http-get:*:image/jpeg:"},
  { "jpg",  UPNP_PHOTO, "http-get:*:image/jpeg:"},
  { "jpe",  UPNP_PHOTO, "http-get:*:image/jpeg:"},
  { "pcd",  UPNP_PHOTO, "http-get:*:image/x-ms-bmp:"},
  { "png",  UPNP_PHOTO, "http-get:*:image/png:"},
  { "pnm",  UPNP_PHOTO, "http-get:*:image/x-portable-anymap:"},
  { "ppm",  UPNP_PHOTO, "http-get:*:image/x-portable-pixmap:"},
  { "qti",  UPNP_PHOTO, "http-get:*:image/x-quicktime:"},
  { "qtf",  UPNP_PHOTO, "http-get:*:image/x-quicktime:"},
  { "qtif", UPNP_PHOTO, "http-get:*:image/x-quicktime:"},
  { "tif",  UPNP_PHOTO, "http-get:*:image/tiff:"},
  { "tiff", UPNP_PHOTO, "http-get:*:image/tiff:"},

  /* Playlist files */
  { "pls", UPNP_PLAYLIST, "http-get:*:audio/x-scpls:"},
  { "m3u", UPNP_PLAYLIST, "http-get:*:audio/mpegurl:"},
  { "asx", UPNP_PLAYLIST, "http-get:*:video/x-ms-asf:"},

  /* Subtitle Text files */
  { "srt", UPNP_TEXT, "http-get:*:text/srt:"}, /* SubRip */
  { "ssa", UPNP_TEXT, "http-get:*:text/ssa:"}, /* SubStation Alpha */
  { "stl", UPNP_TEXT, "http-get:*:text/srt:"}, /* Spruce */
  { "psb", UPNP_TEXT, "http-get:*:text/psb:"}, /* PowerDivX */
  { "pjs", UPNP_TEXT, "http-get:*:text/pjs:"}, /* Phoenix Japanim */
  { "sub", UPNP_TEXT, "http-get:*:text/sub:"}, /* MicroDVD */
  { "idx", UPNP_TEXT, "http-get:*:text/idx:"}, /* VOBsub */
  { "dks", UPNP_TEXT, "http-get:*:text/dks:"}, /* DKS */
  { "scr", UPNP_TEXT, "http-get:*:text/scr:"}, /* MACsub */
  { "tts", UPNP_TEXT, "http-get:*:text/tts:"}, /* TurboTitler */
  { "vsf", UPNP_TEXT, "http-get:*:text/vsf:"}, /* ViPlay */
  { "zeg", UPNP_TEXT, "http-get:*:text/zeg:"}, /* ZeroG */
  { "mpl", UPNP_TEXT, "http-get:*:text/mpl:"}, /* MPL */

  /* Miscellaneous text files */
  { "bup", UPNP_TEXT, "http-get:*:text/bup:"}, /* DVD backup */
  { "ifo", UPNP_TEXT, "http-get:*:text/ifo:"}, /* DVD information */

  { NULL, NULL, NULL}
};

char *mime_get_protocol (struct mime_type_t *mime)
{
  char protocol[512];
  
  if (!mime)
    return NULL;

  sprintf (protocol, "%s", mime->mime_protocol);
  strcat (protocol, "*");
  return strdup (protocol);
}
